File: src/algorithms/Clone_Graph.py
# -*- coding: utf-8 -*-


# Definition for a undirected graph node
class UndirectedGraphNode:
    def __init__(self, x):
        self.label = x
        self.neighbors = []


class Solution:
    def dfs(self, root):
        ret = self.vis[root.label]
        for node in root.neighbors:
            if node.label not in self.vis:
                self.vis[node.label] = UndirectedGraphNode(node.label)
                self.dfs(node)
            ret.neighbors.append(self.vis[node.label])
        return ret

    # @param node, a undirected graph node
    # @return a undirected graph node
    def cloneGraph(self, node):
        if node is None:
            return None
        self.vis = {
            node.label: UndirectedGraphNode(node.label)
        }
        ret = self.dfs(node)
        return ret


if __name__ == "__main__":
    solution = Solution()
    a = UndirectedGraphNode(1)
    b = UndirectedGraphNode(2)
    c = UndirectedGraphNode(3)
    a.neighbors, b.neighbors, c.neighbors = [b], [c], [a]
    node = solution.cloneGraph(a)
    print (
        node.label,
        node.neighbors[-1].label,
        node.neighbors[-1].neighbors[-1].label,
        node.neighbors[-1].neighbors[-1].neighbors[-1].label
    )

File: src/algorithms/cpp/Clone_Graph.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <vector>
#include <map>
#include <set>
using namespace std;
#define out(v) cerr << #v << ": " << (v) << endl
#define SZ(v) ((int)(v).size())
const int maxint = -1u>>1;
template <class T> bool get_max(T& a, const T &b) {return b > a? a = b, 1: 0;}
template <class T> bool get_min(T& a, const T &b) {return b < a? a = b, 1: 0;}

struct UndirectedGraphNode {
    int label;
    vector<UndirectedGraphNode *> neighbors;
    UndirectedGraphNode(int x) : label(x) {};
};

class Solution {
private:
    map <int, UndirectedGraphNode*> mp;

private:
    UndirectedGraphNode* dfs(UndirectedGraphNode *node) {
        UndirectedGraphNode *new_node = new UndirectedGraphNode(node->label);
        mp[node->label] = new_node;
        for (int i = 0; i < SZ(node->neighbors); ++ i) {
            if (mp[node->neighbors[i]->label] == NULL)
                new_node->neighbors.push_back(dfs(node->neighbors[i]));
            else
                new_node->neighbors.push_back(mp[node->neighbors[i]->label]);
        }
        return new_node;
    }

public:
    UndirectedGraphNode* cloneGraph(UndirectedGraphNode *node) {
        if (node == NULL)   return node;
        else                return dfs(node);
    }
};

int main() {
    return 0;
}
